src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerMethodDetails.cs

```diff
@@ -0,0 +1,185 @@
+            {
+                throw new InvalidOperationException($"{handlerType.FullName} is missing {nameof(LanguageServerEndpointAttribute)}");
+            }
+        }
+
+        return (methodAttribute.Method, methodAttribute.Languages);
+
+        static LanguageServerEndpointAttribute? GetMethodAttributeFromHandlerMethod(Type handlerType, Type? requestType, Type contextType, Type? responseType)
+        {
+            const string HandleRequestName = nameof(IRequestHandler<object, object, object>.HandleRequestAsync);
+            const string HandleRequestSuffix = "." + HandleRequestName;
+            const string HandleNotificationName = nameof(INotificationHandler<object, object>.HandleNotificationAsync);
+            const string HandleNotificationSuffix = "." + HandleNotificationName;
+
+            foreach (var methodInfo in handlerType.GetRuntimeMethods())
+            {
+                if (MethodInfoMatches(methodInfo))
+                    return methodInfo.GetCustomAttribute<LanguageServerEndpointAttribute>();
+            }
+
+            throw new InvalidOperationException("Somehow we are missing the method for our registered handler");
+
+            bool MethodInfoMatches(MethodInfo methodInfo)
+            {
+                var methodName = methodInfo.Name;
+
+                switch (requestType != null, responseType != null)
+                {
+                    case (true, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleRequestSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+
+                    case (true, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+                }
+            }
+
+            static bool NameMatches(MethodInfo methodInfo, string name, string suffix)
+            {
+                var methodName = methodInfo.Name;
+
+                return methodName == name || methodName.EndsWith(suffix);
+            }
+
+            static bool TypesMatch(MethodInfo methodInfo, Type?[] types)
+            {
+                var parameters = methodInfo.GetParameters();
+                if (parameters.Length != types.Length)
+                    return false;
+
+                for (int i = 0, n = parameters.Length; i < n; i++)
```

Comment:
curious, why this format instead of `for (int i = 0; i < parameters.Length; i++)`

Reply:
I didn't actually write this code, so I'm not certain why. It was moved from here: https://github.com/dotnet/roslyn/blob/5f778bde9a17c34480dc83f3a17bf4cd9277d630/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerProvider.cs#L178-L182. @CyrusNajmabadi do recall why you wrote the loop this way? I'm kinda curious too. :smile: